### PR TITLE
refactor(permissions): split up Descriptor into Allow, Deny, and Query

### PR DESCRIPTION
**Pull request by dsherret**
This makes the permission system more versatile.

### CODE REVIEW

**Comment by lowlighter** [comment]
Any chance taking a look at #14668 during this refactor ?

**Comment by dsherret** [comment]
@lowlighter no, that would be too many changes for this PR.

File: cli/args/mod.rs [hunk 1]
```diff
@@ -1082,7 +1081,7 @@ impl CliOptions {
             let specifier = specifier.clone();
             async move {
               let file = file_fetcher
-                .fetch(&specifier, &PermissionsContainer::allow_all())
+                .fetch_bypass_permissions(&specifier)
```
Comment (bartlomieju) [inline cli/args/mod.rs:1084]: I see where this is coming from, but for https://github.com/denoland/deno/pull/25469 this will have to be rewritten anyway. I guess we can land it for now though

File: cli/factory.rs [hunk 2]
```diff
@@ -758,11 +782,17 @@ impl CliFactory {
...
+      if cli_options.code_cache_enabled() {
+        Some(self.code_cache()?.clone())
+      } else {
+        None
+      },
+      self.feature_checker()?.clone(),
+      self.fs().clone(),
+      maybe_file_watcher_communicator,
+      self.maybe_inspector_server()?.clone(),
+      cli_options.maybe_lockfile().cloned(),
```
Comment (bartlomieju) [inline cli/factory.rs:795]: Is it necessary to shuffle all these options in this PR? Makes it quite a bit harder to figure out what is actually important that's changed

File: cli/file_fetcher.rs [hunk 3]
```diff
@@ -515,11 +515,29 @@ impl FileFetcher {
...
+    &self,
+    specifier: &ModuleSpecifier,
+  ) -> Result<File, AnyError> {
+    self.fetch_inner(specifier, None).await
+  }
+
+  // todo(THIS PR): lol what... this is never used?
   /// Fetch a source file and asynchronously return it.
+  #[inline(always)]
   pub async fn fetch(
```
Comment (bartlomieju) [inline cli/file_fetcher.rs:529]: This function was used in a lot of places that mostly fetched configuration files and used "all permissions"

File: cli/file_fetcher.rs [hunk 4]
```diff
@@ -583,7 +601,9 @@ impl FileFetcher {
       specifier
     );
     let scheme = get_validated_scheme(specifier)?;
-    options.permissions.check_specifier(specifier)?;
+    if let Some(permissions) = &options.permissions {
+      permissions.check_specifier(specifier)?;
+    }
```
Comment (bartlomieju) [inline cli/file_fetcher.rs:606]: Again for https://github.com/denoland/deno/pull/25469 this will have to change again; maybe don't use `Option<>` here and always do the check?

File: cli/lsp/registries.rs [hunk 5]
```diff
@@ -481,7 +480,7 @@ impl ModuleRegistry {
         file_fetcher
         .fetch_with_options(FetchOptions {
           specifier: &specifier,
-          permissions: &PermissionsContainer::allow_all(),
+          permissions: None,
```
Comment (bartlomieju) [inline cli/lsp/registries.rs:483]: Yeah - using `None` instead of `PermissionsContainer::allow_all()` is a lot less descriptive and then we're gonna have to carry over a list of available hosts when we implement https://github.com/denoland/deno/pull/25469 anyway.

File: ext/fetch/lib.rs [hunk 6]
```diff
@@ -316,12 +321,16 @@ impl FetchPermissions for deno_permissions::PermissionsContainer {
...
   #[inline(always)]
-  fn check_read(
+  fn check_read<'a>(
     &mut self,
-    path: &Path,
+    path: &'a Path,
     api_name: &str,
-  ) -> Result<(), AnyError> {
-    deno_permissions::PermissionsContainer::check_read(self, path, api_name)
+  ) -> Result<Cow<'a, Path>, AnyError> {
```
Comment (bartlomieju) [inline ext/fetch/lib.rs:328]: I don't understand why it needs to have different return type, can you elaborate?

File: runtime/ops/permissions.rs [hunk 7]
```diff
@@ -90,11 +142,33 @@ pub fn op_revoke_permission(
   state: &mut OpState,
   #[serde] args: PermissionArgs,
 ) -> Result<PermissionStatus, AnyError> {
-  let mut permissions = state.borrow_mut::<PermissionsContainer>().0.lock();
+  // todo(dsherret): don't have this function use the properties of
+  // permission container
+  let permissions_container = state.borrow_mut::<PermissionsContainer>();
```
Comment (bartlomieju) [inline runtime/ops/permissions.rs:147]: What's wrong with that?

File: runtime/permissions/lib.rs [hunk 8]
```diff
@@ -324,27 +325,44 @@ pub trait Descriptor: Eq + Clone + Hash {
     api_name: Option<&str>,
   ) -> Result<(), AnyError>;
 
-  fn flag_name() -> &'static str;
-  fn name(&self) -> Cow<str>;
-  // By default, specifies no-stronger-than relationship.
-  // As this is not strict, it's only true when descriptors are the same.
-  fn stronger_than(&self, other: &Self) -> bool {
-    self == other
+  /// Gets if this query descriptor should revoke the provided allow descriptor.
```
Comment (bartlomieju) [inline runtime/permissions/lib.rs:328]:
```suggestion
  /// Gets called if this query descriptor should revoke the provided allow descriptor.
```

File: runtime/ops/permissions.rs (hunk 7 above)
Comment (dsherret) [reply runtime/ops/permissions.rs:147]: It's exposing a mutex, it's having code that shouldn't be done/exposed at this level (parsing descriptors), and it's coupling the private api to this part of the code.

File: ext/fetch/lib.rs (hunk 6 above)
Comment (dsherret) [reply ext/fetch/lib.rs:328]: I pushed an update that adds a description to the must_use attribute. It's to mitigate TOCTOU attacks.

File: cli/file_fetcher.rs (hunk 3 above)
Comment (dsherret) [reply cli/file_fetcher.rs:530]: Yeah, it just seems we always fetched with all permissions and effectively didn't use what is now this code.

File: cli/factory.rs (hunk 2 above)
Comment (dsherret) [reply cli/factory.rs:795]: I'll do it in a separate PR.

Comment (dsherret) [reply cli/factory.rs:795]: Actually, it seems too small of a change. The only thing that changed here was I added the `self.permission_desc_parser()?.clone(),` argument.

File: cli/lsp/registries.rs (hunk 5 above)
Comment (dsherret) [reply cli/lsp/registries.rs]: I updated this to use a new enum here that has the permissions container in one of the variants and allow all in the other to make this clear.

File: tests/unit/worker_test.ts [hunk 9]
```diff
@@ -482,8 +515,10 @@ Deno.test("Worker limit children permissions granularly", async function () {
     readAbsent: "prompt",
     runGlobal: "prompt",
     runFoo: "granted",
+    runFooPath: "granted",
     runBar: "granted",
     runBaz: "granted",
+    runUnresolved: "prompt", // unresolved binaries remain as "prompt"
```
Comment (dsherret) [inline tests/unit/worker_test.ts:521]: This shows a breaking change--any command name that can't be resolved when launching the worker will be excluded from having allow permissions.

File: cli/file_fetcher.rs [hunk 10]
```diff
@@ -161,9 +161,15 @@ fn get_validated_scheme(
   }
 }
 
+#[derive(Debug, Copy, Clone)]
+pub enum FetchPermissionsOptionRef<'a> {
```
Comment (bartlomieju) [inline cli/file_fetcher.rs:165]: Shouldn't this one be in `cli/cache/mod.rs`? It's a bit concerning that these are spread across various modules. Maybe add `cli/permissions.rs` to colocate all this code?

File: cli/file_fetcher.rs [hunk 11]
```diff
@@ -515,11 +521,33 @@ impl FileFetcher {
...
+    &self,
+    specifier: &ModuleSpecifier,
+  ) -> Result<File, AnyError> {
+    self
+      .fetch_inner(specifier, FetchPermissionsOptionRef::AllowAll)
+      .await
+  }
+
   /// Fetch a source file and asynchronously return it.
+  #[allow(dead_code)] // todo(25469): undo when merging
```
Comment (bartlomieju) [inline cli/file_fetcher.rs:535]: Linking back to https://github.com/denoland/deno/pull/25469

File: cli/standalone/mod.rs [hunk 12]
```diff
@@ -660,24 +661,28 @@ pub async fn run(
   };
 
   let permissions = {
-    let mut permissions =
-      metadata.permissions.to_options(maybe_cwd.as_deref())?;
+    let mut permissions = metadata.permissions.to_options()?;
```
Comment (bartlomieju) [inline cli/standalone/mod.rs:664]: I'm not sure I'm a fan of not explicitly passing CWD - now it becomes buried very deeply

File: ext/fetch/lib.rs [hunk 13]
```diff
@@ -299,10 +299,15 @@ impl Drop for ResourceToBodyAdapter {
 pub trait FetchPermissions {
   fn check_net_url(
     &mut self,
-    _url: &Url,
+    url: &Url,
     api_name: &str,
   ) -> Result<(), AnyError>;
-  fn check_read(&mut self, _p: &Path, api_name: &str) -> Result<(), AnyError>;
+  #[must_use = "the resolved return value to mitigate time-of-check to time-of-use issues"]
```
Comment (bartlomieju) [inline ext/fetch/lib.rs:305]: This is great 👍

File: ext/fetch/lib.rs [hunk 14]
```diff
@@ -316,12 +321,16 @@ impl FetchPermissions for deno_permissions::PermissionsContainer {
   }
 
   #[inline(always)]
-  fn check_read(
+  fn check_read<'a>(
```
Comment (bartlomieju) [inline ext/fetch/lib.rs:324]: Add `must_use` attr as well?

File: runtime/ops/permissions.rs (hunk 7 above)
Comment (bartlomieju) [reply runtime/ops/permissions.rs:139]: Open an issue to keep track of it? Or are you planning to address it before landing?

File: runtime/ops/process.rs [hunk 15]
```diff
@@ -530,7 +531,7 @@ fn compute_run_cmd_and_check_permissions(
 }
 
 struct RunEnv {
-  envs: HashMap<String, String>,
+  envs: HashMap<OsString, OsString>,
```
Comment (bartlomieju) [inline runtime/ops/process.rs:534]: Is this a perf optimization? What's the reason to use `OsString` here?

File: runtime/web_worker.rs [hunk 16]
```diff
@@ -356,6 +357,7 @@ pub struct WebWorker {
 }
 
 pub struct WebWorkerOptions {
+  // todo(dsherret): extract out the service structs from this options bag
```
Comment (bartlomieju) [inline runtime/web_worker.rs:360]: Please, open a tracking issue before landing

File: runtime/worker.rs [hunk 17]
```diff
@@ -201,13 +204,16 @@ pub struct WorkerOptions {
   pub v8_code_cache: Option<Arc<dyn CodeCache>>,
 }
 
+// todo(dsherret): this is error prone to use. We should separate
+// out the WorkerOptions from the services.
```
Comment (bartlomieju) [inline runtime/worker.rs:208]: Ditto

**Review by bartlomieju — COMMENTED** [review]
I want to give `runtime/permissions/lib.rs` one more pass tomorrow

File: runtime/ops/process.rs (hunk 15 above)
Comment (dsherret) [reply runtime/ops/process.rs:534]: It's to keep them in their original encoding. I think it was a bug I introduced earlier to not use `OsString`

File: cli/standalone/mod.rs (hunk 12 above)
Comment (dsherret) [reply cli/standalone/mod.rs:664]: The CWD is not used internally here.

File: cli/file_fetcher.rs (hunk 10 above)
Comment (dsherret) [reply cli/file_fetcher.rs:165]: I don't think so because this is the lowest level this is used at.

Comment (dsherret) [reply cli/file_fetcher.rs:165]: Ah, I see what you mean because there is `FetchPermissionsOption` and the `Ref` version. I'll just put them both here.

File: ext/fetch/lib.rs (hunk 14 above)
Comment (dsherret) [reply ext/fetch/lib.rs:324]: It's on the trait. I think it doesn't need to be here as well?

File: runtime/web_worker.rs (hunk 16 above)
Comment (dsherret) [reply runtime/web_worker.rs:360]: https://github.com/denoland/deno/issues/25634

File: runtime/permissions/lib.rs [hunk 18]
```diff
@@ -310,12 +308,15 @@ impl AsRef<str> for EnvVarName {
   }
 }
 
-pub trait Descriptor: Eq + Clone + Hash {
-  type Arg: From<String>;
+pub trait QueryDescriptor: Debug {
+  type AllowDesc: AllowDescriptor<Query = Self>;
+  type DenyDesc: DenyDescriptor<Query = Self>;
```
Comment (bartlomieju) [inline runtime/permissions/lib.rs:313]: This is a bit confusing - effectively there are circular dependencies between these three types. Do you see any chance we could decouple that with a helper type?

Comment (bartlomieju) [reply runtime/permissions/lib.rs:313]: Also could you add some docstrings around these traits?

File: runtime/permissions/lib.rs [hunk 19]
```diff
@@ -509,55 +527,66 @@ impl<T: Descriptor + Hash> UnaryPermission<T> {
     self.query_desc(desc, AllowPartial::TreatAsPartialGranted)
   }
 
-  fn is_granted(&self, desc: Option<&T>) -> bool {
-    Self::list_contains(desc, self.granted_global, &self.granted_list)
+  fn is_granted(&self, desc: Option<&TQuery>) -> bool {
+    match desc {
+      Some(desc) => {
+        self.granted_global || self.granted_list.iter().any(|v| v.matches(desc))
```
Comment (bartlomieju) [inline runtime/permissions/lib.rs:533]: Why are we using `.matches()` instead of `.stronger_than()` now?

File: runtime/permissions/lib.rs [hunk 20]
```diff
@@ -869,138 +1069,317 @@ impl AsRef<str> for EnvDescriptor {
...
+pub enum RunQueryDescriptor {
+  Path {
+    requested: String,
+    resolved: PathBuf,
+  },
+  /// This variant can't be used to grant permissions. It's mostly
+  /// used so that prompts and everything works the same way as when
+  /// the command is resolved, meaning that a script can't tell
+  /// if a command is resolved or not based on how long something
+  /// takes to ask for permissions.
```
Comment (bartlomieju) [inline runtime/permissions/lib.rs:1081]: Could you reword this docstring? I'm having trouble understanding it - what does "This variant can't be used to grant permissions" mean?

File: runtime/permissions/lib.rs [hunk 21]
```diff
@@ -1607,149 +2018,260 @@ impl Permissions {
 /// case might need to be mutated). Also for the Web Worker API we need a way
 /// to send permissions to a new thread.
 #[derive(Clone, Debug)]
-pub struct PermissionsContainer(pub Arc<Mutex<Permissions>>);
+pub struct PermissionsContainer {
+  // todo(dsherret): make both of these private as the functionality
+  // can just be methods on PermissionsContainer. Additionally, a separate
+  // struct should be created in here that handles creating child permissions
+  // so that the code is not so verbose elsewhere.
+  pub descriptor_parser: Arc<dyn PermissionDescriptorParser>,
```
Comment (bartlomieju) [inline runtime/permissions/lib.rs:2026]: Can you open an issue about this one?

File: runtime/ops/permissions.rs (hunk 7 above)
Comment (dsherret) [reply runtime/ops/permissions.rs:139]: https://github.com/denoland/deno/issues/25634

File: runtime/permissions/lib.rs (hunk 21 above)
Comment (dsherret) [reply runtime/permissions/lib.rs:2026]: I put it in.

File: runtime/permissions/lib.rs (hunk 19 above)
Comment (dsherret) [reply runtime/permissions/lib.rs]: There are multiple specific methods. I've grouped them all on query now.

File: runtime/permissions/lib.rs [hunk 22]
```diff
@@ -324,27 +325,46 @@ pub trait Descriptor: Eq + Clone + Hash {
...
+pub trait AllowDescriptor: Debug + Eq + Clone + Hash {
+  type Query: QueryDescriptor;
+
+  fn as_query(&self) -> Self::Query;
+}
+
+pub trait DenyDescriptor: Debug + Eq + Clone + Hash {
+  type Query: QueryDescriptor;
+
+  fn is_partial(&self, query: &Self::Query) -> bool;
```
Comment (dsherret) [inline runtime/permissions/lib.rs:353]: I'll look into shifting this one over to query as well (ran out of battery where I am)

File: runtime/permissions/lib.rs (hunk 18 above)
Comment (dsherret) [reply runtime/permissions/lib.rs]: I got rid of the additional traits and now there's only a QueryDescriptor.

File: runtime/permissions/lib.rs (hunk 20 above)
Comment (dsherret) [reply runtime/permissions/lib.rs]: I updated it a bit. The variant gets created when the command path can't be resolved, so it will prompt the users for permissions about it, but it doesn't actually do anything when the user grants permissions. It's to prevent someone trying to probe for permissions to find out what's on the path.

File: tests/unit/os_test.ts [hunk 23]
```diff
@@ -79,7 +79,9 @@ Deno.test(
     ) => {
       const src = `
       console.log(
-        ${JSON.stringify(Object.keys(expectedEnv))}.map(k => Deno.env.get(k))
+        ${
+        JSON.stringify(Object.keys(expectedEnv))
+      }.map(k => Deno.env.get(k) ?? null)
```
Comment (satyarohith) [inline tests/unit/os_test.ts:84]: why is this change required? I assume it's unrelated?

Comment (dsherret) [reply tests/unit/os_test.ts:84]: `undefined` isn't valid JSON, but `null` is (it does a `JSON.parse` below). I ran into this issue while developing where I introduced a bug, then it was `undefined`, then the error message was not great.

**Comment by dsherret** [comment]
> From what I can tell, we do a lot more canonicalization for FS ops now (maybe including more getcwd calls?) 

It should be doing the same or less now with --allow-all. It doesn't canonicalize like before.

**Comment by lucacasonato** [comment]
I meant without --allow-all

**Comment by dsherret** [comment]
Yeah should be the same without allow-run.

File: runtime/permissions/lib.rs [hunk 24]
```diff
@@ -1486,57 +1747,140 @@ impl Permissions {
...
+              None // convert to no permissions if now empty
+            } else {
+              Some(Ok(resolved_allow_run))
+            }
+          }
+          Err(err) => Some(Err(err)),
+        }
+      })
+      .transpose()?;
+    // add the allow_run list to deno_write
```
[the remark this answers is not in the document]
Comment (dsherret) [reply runtime/permissions/lib.rs:1814]: Not sure. It was here before: https://github.com/denoland/deno/blob/eb8ee95f08186c948e5b83501cedd59d6e3b4ef2/cli/args/flags.rs#L733. I'll remove it.

[the remark this answers is not in the document]
Comment (dsherret) [reply runtime/permissions/lib.rs:1814]: Oh wait, I wrote this lol. Should be deny-write.

**Comment by dsherret** [comment]
Some benchmarks:

```
$ cat main.ts
for (var i = 0; i < 100_000; i++) {
  const text = Deno.readTextFileSync("./main.ts");
  console.log(text.length);
}
$ hyperfine --warmup=3 "./deno_main run --allow-read=main.ts main.ts" "./deno_new run --allow-read=main.ts main.ts"
Benchmark 1: ./deno_main run --allow-read=main.ts main.ts
  Time (mean ± σ):      1.955 s ±  0.018 s    [User: 0.513 s, System: 1.446 s]
  Range (min … max):    1.931 s …  1.989 s    10 runs
 
Benchmark 2: ./deno_new run --allow-read=main.ts main.ts
  Time (mean ± σ):      1.932 s ±  0.011 s    [User: 0.495 s, System: 1.441 s]
  Range (min … max):    1.920 s …  1.953 s    10 runs
 
Summary
  ./deno_new run --allow-read=main.ts main.ts ran
    1.01 ± 0.01 times faster than ./deno_main run --allow-read=main.ts main.ts
```

**Review by bartlomieju — APPROVED** [review]
LGTM
